feat(commands): Add support for `Range[LargeInt, ...]`

Enegg opened the pull request:
## Summary
Closes #787
This PR adds support for defining a range of accepted values for `commands.LargeInt`.

If both bounds are specified, we restrict the length of the string input according to the number of digits (and sign) of the bounds.

We also raise an error during conversion, if the provided value would exceed the defined range (for example `Range[LargeInt, 100, 200]` would have actual possible input of 100-999).

I've also implemented a check in case the developer attempts to define a range exceeding ±2**53 with the normal `int` type, directing them to `LargeInt`.

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `pdm lint`
    - [x] I have type-checked the code by running `pdm pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

Enegg committed:
Add support for `Range[LargeInt, ...]`

Enegg committed:
Use wider type ignore

Enegg committed:
Fix typo: =< → <=

DJStompZone commented:
<p align="center"><img width="40%" src="https://github.com/DisnakeDev/disnake/assets/85457381/1aa546a1-c4a4-4b6a-a9b4-94795e3d48b8"></p>

Enegg commented:
Tangentially related: we could add slice support for defining ranges (`Range[int, 0:32]`, `String[str, :10]`). Not sure how useful it'd be.

Enegg committed:
Fix 3.8/3.9 ellipsis issues

Enegg committed:
Set string length for certain ranges with one unbound

Enegg committed:
Changelog

Enegg committed:
Fix test regex

Enegg committed:
Oh no, missing newline!

Enegg commented on disnake/ext/commands/params.py:
```diff
@@ -717,18 +757,36 @@ def parse_annotation(self, annotation: Any, converter_mode: bool = False) -> boo
+                self.min_length = _int_to_str_len(self.min_value)  # pyright: ignore
+
+            elif self.max_value is not None and self.max_value < 0:
+                # -inf == max_value <= min_value < 0
```
Should be min <= max 😔

Enegg committed:
Extract function

onerandomusername commented:
Hi, would you please resolve conflicts?

Enegg committed:
Thanks codemod

Enegg committed:
Fix 3.9 non-existent Union type

Enegg committed:
Clarify changelog

shiftinv commented on disnake/ext/commands/errors.py:
```diff
@@ -572,6 +572,37 @@ def __init__(self, argument: str) -> None:
+
+    This inherits from :exc:`BadArgument`
+
+    .. versionadded:: 2.11
```
```suggestion
    .. versionadded:: |vnext|
```

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -155,7 +147,47 @@ def _xt_to_xe(xe: Optional[float], xt: Optional[float], direction: float = 1) ->
+    # both bounds negative: len(str(min_value)) >= len(str(max_value))
+    if opposite_sign:
+        return 1, max(min_, max_)
+    return min(min_, max_), max(min_, max_)
```
I haven't looked into it, but assuming we can expect `min_value < max_value` to always be true here, I believe this could be simplified a little

Enegg committed:
Merge branch 'master' into Eneg/large-range

Enegg committed:
Apply vi's suggestions

Co-authored-by: vi <8530778+shiftinv@users.noreply.github.com>
Signed-off-by: Eneg <42005170+Enegg@users.noreply.github.com>

Enegg committed:
Implement other suggestions

Enegg committed:
Move Range/String out of TYPE_CHECKING block

Enegg commented on disnake/ext/commands/errors.py:
```diff
@@ -572,6 +572,37 @@ def __init__(self, argument: str) -> None:
         super().__init__(f"{argument} is not able to be converted to an integer")
 
 
+class LargeIntOutOfRange(BadArgument):
```
https://github.com/DisnakeDev/disnake/pull/1201/commits/009da96e355a1dd626062b402797b156a6e2dee4

Enegg commented on disnake/ext/commands/params.py:
```diff
@@ -155,7 +147,47 @@ def _xt_to_xe(xe: Optional[float], xt: Optional[float], direction: float = 1) ->
+        # -inf == min_value <= max_value < 0
+        return None, _int_to_str_len(max_value)
+
+    return None, None
```
https://github.com/DisnakeDev/disnake/pull/1201/commits/009da96e355a1dd626062b402797b156a6e2dee4

Enegg committed:
Create a docs section about Large Integers, and link to it

Enegg committed:
Duplicate Sequence import

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -375,18 +413,27 @@ def __post_init__(self) -> None:
                     continue
 
-                if self.underlying_type is int and not isinstance(value, int):
+                if self.underlying_type is not float and not isinstance(value, int):
```
Hmm, fair. Perhaps `self.underlying_type is not float and isinstance(value, float)`?

Enegg committed:
Refactor condition with two negations

Enegg committed:
Merge branch 'Eneg/large-range' of https://github.com/Enegg/disnake into Eneg/large-range

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -154,7 +150,47 @@ def _xt_to_xe(xe: float | None, xt: float | None, direction: float = 1) -> float
+    # both bounds negative: len(str(min_value)) >= len(str(max_value))
+    if opposite_sign:
+        return 1, max(min_, max_)
+    return min(min_, max_), max(min_, max_)
```
```suggestion
    min_len = _int_to_str_len(min_value)
    max_len = _int_to_str_len(max_value)
    opposite_sign = (min_value < 0) ^ (max_value < 0)
    # both bounds positive: len(str(min_value)) <= len(str(max_value))
    # smaller bound negative: the range includes 0, which sets the minimum length to 1
    # both bounds negative: len(str(min_value)) >= len(str(max_value))
    if opposite_sign:
        return 1, max(min_len, max_len)
    return min(min_len, max_len), max(min_len, max_len)
```

Enegg replied:
Hmm, this can still be confusing, as `min_len` isn't *literally* the minimum input length; rather, it's the digit count (and sign) of the minimum input value

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -154,7 +150,47 @@ def _xt_to_xe(xe: float | None, xt: float | None, direction: float = 1) -> float
+
+    elif max_value is not None and max_value < 0:
+        # -inf == min_value <= max_value < 0
+        return 1, _int_to_str_len(max_value)
```
I think my brain might be breaking from thinking about this for too long, but - is this correct? The string length shouldn't have an upper bound, since it can go infinitely far towards -inf, no?
(I'll add a couple test cases later to sanity check what we'd expect from these functions)

Enegg replied:
Hmm, I think your long thinking is justified, you are correct that this is wrong

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -335,89 +379,105 @@ def __repr__(self) -> str:
-        _allowed_types = (int, float)
+            if self.underlying_type is int and abs(value) >= 2**53 - 1:
+                msg = (
+                    "Discord has upper input limit on integer input type of ±2**53.\n"
```
```suggestion
                    "Discord imposes an upper input limit of ±2**53 on integer input types.\n"
```

shiftinv committed:
Merge branch 'master' into Eneg/large-range

Signed-off-by: vi <8530778+shiftinv@users.noreply.github.com>

shiftinv committed:
fix(docs): document `Range` and `String` aliases as their original classes

shiftinv committed:
fix(docs): don't autouse `classalias` documenter

This isn't immediately an issue, but prevents this documenter from being
used for members that are classes (for whatever reason that would be the case).
We only want to use it if explicitly specified through `autoclassalias::`

shiftinv commented on tests/ext/commands/test_params.py:
(case 7 and 8 currently fail)
```suggestion

    @pytest.mark.parametrize(
        ("bounds", "expected_len"),
        # ((min_value, max_value), (expected_min_length, expected_max_length))
        [
            # positive only
            ((0, ...), (1, None)),
            ((1, ...), (1, None)),
            ((10, ...), (2, None)),
            ((5000, ...), (4, None)),
            ((1, 5), (1, 1)),
            ((1, 10000), (1, 5)),
            # negative only
            ((..., -1), (2, None)),
            ((..., -123), (4, None)),
            ((-10, -1), (2, 3)),
            # crossing zero boundary (min length should always be exactly 1)
            ((-1, ...), (1, None)),
            ((..., 50), (1, None)),
            ((-1, 1), (1, 2)),
            ((-1, 100), (1, 3)),
            ((-100, 50), (1, 4)),
        ],
        ids=repr,
    )
    def test_large_range(self, bounds, expected_len) -> None:
        info = commands.ParamInfo()
        info.parse_annotation(commands.Range[(commands.LargeInt, *bounds)])

        assert info.large
        assert (info.min_length, info.max_length) == expected_len

```

Enegg committed:
Apply suggestions from code review

Co-authored-by: vi <8530778+shiftinv@users.noreply.github.com>
Signed-off-by: Eneg <42005170+Enegg@users.noreply.github.com>

Enegg committed:
Fix -inf == min_value <= max_value < 0
Add extra tests

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -252,22 +286,36 @@ def decorator(func: FuncT) -> FuncT:
+        # the classes are defined as private, but have public aliases.
+        # we don't want the users using them via the private names,
+        # since those conflict with type checkers
+        name = cls.__name__ = cls.__name__.removeprefix("_")
```
Is modifying `cls.__name__` here intentional?

Enegg replied:
Yes, I think we want the ranges to appear in repr etc. exactly as if the name wasn't `_` prefixed
I haven't double checked if this is actually reflected by the repr though

shiftinv replied:
Ah, I see. Feels a bit cursed, but should be fine, then c:
(and yeah, it does show up correctly in the repr)

shiftinv commented on disnake/ext/commands/params.py:
```diff
@@ -155,7 +147,47 @@ def _xt_to_xe(xe: Optional[float], xt: Optional[float], direction: float = 1) ->
 
-class Injection(Generic[P, T_]):
+def _int_to_str_len(number: int) -> int:
+    """Returns `len(str(number))`, i.e. character count of base 10 signed repr of `number`."""
```
While a couple dozen nanoseconds perhaps aren't all that important on a cold path like this, I think this can be simplified to just `len(str(x))` nonetheless

shiftinv committed:
Merge branch 'master' into Eneg/large-range

shiftinv reviewed (approved):
lgtm! Thank you for all your work on this feature (and patience), really appreciate it c:
